pkg/sdkserver/routes.go

```diff
@@ -0,0 +1,231 @@
+}
+
+// listTools will return the output of `gptscript --list-tools`
+func (s *server) listTools(w http.ResponseWriter, r *http.Request) {
+	logger := gcontext.GetLogger(r.Context())
+	var prg types.Program
+	if r.ContentLength != 0 {
+		reqObject := new(toolOrFileRequest)
+		err := json.NewDecoder(r.Body).Decode(reqObject)
+		if err != nil {
+			writeError(logger, w, http.StatusBadRequest, fmt.Errorf("failed to decode request body: %w", err))
+			return
+		}
+
+		if reqObject.Content != "" {
+			prg, err = loader.ProgramFromSource(r.Context(), reqObject.Content, reqObject.SubTool, loader.Options{Cache: s.client.Cache})
+		} else if reqObject.File != "" {
+			prg, err = loader.Program(r.Context(), reqObject.File, reqObject.SubTool, loader.Options{Cache: s.client.Cache})
+		} else {
+			prg, err = loader.ProgramFromSource(r.Context(), reqObject.ToolDef.String(), reqObject.SubTool, loader.Options{Cache: s.client.Cache})
+		}
+		if err != nil {
+			writeError(logger, w, http.StatusInternalServerError, fmt.Errorf("failed to load program: %w", err))
+			return
+		}
+	}
+
+	tools := s.client.ListTools(r.Context(), prg)
+	sort.Slice(tools, func(i, j int) bool {
+		return tools[i].Name < tools[j].Name
+	})
+
+	lines := make([]string, 0, len(tools))
+	for _, tool := range tools {
+		if tool.Name == "" {
+			tool.Name = prg.Name
+		}
+
+		// Don't print instructions
+		tool.Instructions = ""
+
+		lines = append(lines, tool.String())
+	}
+
+	writeResponse(logger, w, map[string]any{"stdout": strings.Join(lines, "\n---\n")})
+}
+
+// listModels will return the output of `gptscript --list-models`
+func (s *server) listModels(w http.ResponseWriter, r *http.Request) {
+	logger := gcontext.GetLogger(r.Context())
+	var providers []string
+	if r.ContentLength != 0 {
+		reqObject := new(modelsRequest)
+		if err := json.NewDecoder(r.Body).Decode(reqObject); err != nil {
+			writeError(logger, w, http.StatusBadRequest, fmt.Errorf("failed to decode request body: %w", err))
+			return
+		}
+
+		providers = reqObject.Providers
+	}
+
+	out, err := s.client.ListModels(r.Context(), providers...)
+	if err != nil {
+		writeError(logger, w, http.StatusInternalServerError, fmt.Errorf("failed to list models: %w", err))
+		return
+	}
+
+	writeResponse(logger, w, map[string]any{"stdout": strings.Join(out, "\n")})
+}
+
+// execHandler is a general handler for executing tools with gptscript. This is mainly responsible for parsing the request body.
+// Then the options and tool are passed to the process function.
+func (s *server) execHandler(w http.ResponseWriter, r *http.Request) {
+	logger := gcontext.GetLogger(r.Context())
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		writeError(logger, w, http.StatusInternalServerError, fmt.Errorf("failed to read request body: %w", err))
+		return
+	}
+
+	if body[len(body)-1] != '}' {
+		body = append(body, '"', '}')
+	}
```

Comment:
Is this in case a model making calls through the SDK fails to finish a JSON string?

Reply:
This was supposed to be removed.